In [1]:
# Import necessary libraries and modules
import os
import sys
import pandas as pd

sys.path.insert(0, os.path.dirname(os.getcwd()))
from scripts.logger import Logger

# Suppress FutureWarnings
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [2]:
# Initialize logger
logger = Logger('rossmann_analysis.log')

In [3]:
#  Load Data
logger.log('Loading data...')
train_data = pd.read_csv('../data/train.csv', low_memory=False, index_col=False)
test_data = pd.read_csv('../data/test.csv', low_memory= False, index_col=False)
store_data = pd.read_csv('../data/store.csv', low_memory= False, index_col=False)

logger.log('Data loaded successfully.')

In [4]:
logger.log("First 5 rows of the training data (train_data)")
train_data.head()

Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday  
0              1  
1              1  
2              1  
3              1  
4              1

In [7]:
logger.log("First 5 rows of the testing data (test_data)")
test_data.head()

2024-09-20 16:49:03,138 - INFO - First 5 rows of the testing data (test_data)


Id  Store  DayOfWeek        Date  Open  Promo StateHoliday  SchoolHoliday
0   1      1          4  2015-09-17   1.0      1            0              0
1   2      3          4  2015-09-17   1.0      1            0              0
2   3      7          4  2015-09-17   1.0      1            0              0
3   4      8          4  2015-09-17   1.0      1            0              0
4   5      9          4  2015-09-17   1.0      1            0              0

In [8]:
logger.log("First 5 rows of the store data (store_data)")
store_data.head()

2024-09-20 16:49:35,534 - INFO - First 5 rows of the store data (store_data)


Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                        9.0   
1      2         a          a                570.0                       11.0   
2      3         a          a              14130.0                       12.0   
3      4         c          c                620.0                        9.0   
4      5         a          a              29910.0                        4.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              NaN              NaN   
1                    2007.0       1             13.0           2010.0   
2                    2006.0       1             14.0           2011.0   
3                    2009.0       0              NaN              NaN   
4                    2015.0       0              NaN              NaN   

     PromoInterval  
0              NaN  
1  Jan,Apr,Jul,Oct  
2  Jan,Apr,Jul,Oct  
3              NaN  
4              NaN

### Statistical analysis

In [11]:
logger.log("train_data.info()")
train_data.info()

2024-09-20 17:02:31,570 - INFO - train_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   Store          1017209 non-null  int64 
 1   DayOfWeek      1017209 non-null  int64 
 2   Date           1017209 non-null  object
 3   Sales          1017209 non-null  int64 
 4   Customers      1017209 non-null  int64 
 5   Open           1017209 non-null  int64 
 6   Promo          1017209 non-null  int64 
 7   StateHoliday   1017209 non-null  object
 8   SchoolHoliday  1017209 non-null  int64 
dtypes: int64(7), object(2)
memory usage: 69.8+ MB


In [12]:
logger.log("test_data.info()")
test_data.info()

2024-09-20 17:03:46,851 - INFO - test_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41088 entries, 0 to 41087
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             41088 non-null  int64  
 1   Store          41088 non-null  int64  
 2   DayOfWeek      41088 non-null  int64  
 3   Date           41088 non-null  object 
 4   Open           41077 non-null  float64
 5   Promo          41088 non-null  int64  
 6   StateHoliday   41088 non-null  object 
 7   SchoolHoliday  41088 non-null  int64  
dtypes: float64(1), int64(5), object(2)
memory usage: 2.5+ MB


In [13]:
logger.log("store_data.info()")
store_data.info()

2024-09-20 17:03:53,839 - INFO - store_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1115 entries, 0 to 1114
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Store                      1115 non-null   int64  
 1   StoreType                  1115 non-null   object 
 2   Assortment                 1115 non-null   object 
 3   CompetitionDistance        1112 non-null   float64
 4   CompetitionOpenSinceMonth  761 non-null    float64
 5   CompetitionOpenSinceYear   761 non-null    float64
 6   Promo2                     1115 non-null   int64  
 7   Promo2SinceWeek            571 non-null    float64
 8   Promo2SinceYear            571 non-null    float64
 9   PromoInterval              571 non-null    object 
dtypes: float64(5), int64(2), object(3)
memory usage: 87.2+ KB


In [14]:
logger.log("Shape of data")
print ('shape of train data:' ,train_data.shape)
print ('shape of test data:' ,test_data.shape)
print('shape of store data:',store_data.shape)

2024-09-20 17:13:40,267 - INFO - Shape of data


shape of train data: (1017209, 9)
shape of test data: (41088, 8)
shape of store data: (1115, 10)


### Data Overview

- **Train Data Shape:** `((1017209, 9))`
- **Test Data Shape:** `(41088, 8)`
- **Store Data Shape:** `(1115, 10)`

#### Train Data Info
- Total Entries: `1,017,209`
- Columns: `9` (Store, DayOfWeek, Date, Sales, Customers, Open, Promo, StateHoliday, SchoolHoliday)
- Data Types: 
  - `int64`: 7 columns
  - `object`: 2 columns
- Non-Null Entries:
  - Most columns are complete except for the **Open** column, which has 11 missing values.

#### Test Data Info:
- Total Entries: `41,088`
- Columns: `8` (Id, Store, DayOfWeek, Date, Open, Promo, StateHoliday, SchoolHoliday)
- Data Types: 
  - `int64`: 5 columns
  - `float64`: 1 column
  - `object`: 2 columns
- Non-Null Entries:
  - all columns are complete, no missing values.

#### Store Data Info:
- Total Entries: `1,115`
- Columns: `10` (Store, StoreType, Assortment, CompetitionDistance, CompetitionOpenSinceMonth, etc.)
- Data Types:
  - `int64`: 2 columns
  - `float64`: 5 columns
  - `object`: 3 columns
- Missing Values:
  - **CompetitionDistance**: 3 missing values
  - **CompetitionOpenSinceMonth** & **Year**: Significant missing values (~350 entries)
  - **Promo2SinceWeek** & **Promo2SinceYear**: Around 544 missing entries


In [8]:
logger.log("Merge train and test data with store data")
train_df = pd.merge(train_data, store_data, on='Store', how='left')
test_df = pd.merge(test_data, store_data, on='Store', how='left')

2024-09-20 17:52:42,660 - INFO - Merge train and test data with store data


In [7]:
logger.log("First 5 rows of the merged train data (train_df)")
train_df.head()

2024-09-20 17:51:06,483 - INFO - First 5 rows of the merged train data (train_df)


Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday StoreType Assortment  CompetitionDistance  \
0              1         c          a               1270.0   
1              1         a          a                570.0   
2              1         a          a              14130.0   
3              1         c          c                620.0   
4              1         a          a              29910.0   

   CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                        9.0                    2008.0       0   
1                       11.0                    2007.0       1   
2                       12.0                    2006.0       1   
3                        9.0                    2009.0       0   
4                        4.0                    2015.0       0   

   Promo2SinceWeek  Promo2SinceYear    PromoInterval  
0              NaN              NaN              NaN  
1             13.0           2010.0  Jan,Apr,Jul,Oct  
2             14.0           2011.0  Jan,Apr,Jul,Oct  
3              NaN              NaN              NaN  
4              NaN              NaN              NaN

In [9]:
logger.log("First 5 rows of the merged test data (test_df)")
test_df.head()

2024-09-20 17:53:51,827 - INFO - First 5 rows of the merged test data (test_df)


Id  Store  DayOfWeek        Date  Open  Promo StateHoliday  SchoolHoliday  \
0   1      1          4  2015-09-17   1.0      1            0              0   
1   2      3          4  2015-09-17   1.0      1            0              0   
2   3      7          4  2015-09-17   1.0      1            0              0   
3   4      8          4  2015-09-17   1.0      1            0              0   
4   5      9          4  2015-09-17   1.0      1            0              0   

  StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0         c          a               1270.0                        9.0   
1         a          a              14130.0                       12.0   
2         a          c              24000.0                        4.0   
3         a          a               7520.0                       10.0   
4         a          c               2030.0                        8.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              NaN              NaN   
1                    2006.0       1             14.0           2011.0   
2                    2013.0       0              NaN              NaN   
3                    2014.0       0              NaN              NaN   
4                    2000.0       0              NaN              NaN   

     PromoInterval  
0              NaN  
1  Jan,Apr,Jul,Oct  
2              NaN  
3              NaN  
4              NaN

In [10]:
logger.log("train_df.info()")
train_df.info()

2024-09-20 17:56:27,771 - INFO - train_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 18 columns):
 #   Column                     Non-Null Count    Dtype  
---  ------                     --------------    -----  
 0   Store                      1017209 non-null  int64  
 1   DayOfWeek                  1017209 non-null  int64  
 2   Date                       1017209 non-null  object 
 3   Sales                      1017209 non-null  int64  
 4   Customers                  1017209 non-null  int64  
 5   Open                       1017209 non-null  int64  
 6   Promo                      1017209 non-null  int64  
 7   StateHoliday               1017209 non-null  object 
 8   SchoolHoliday              1017209 non-null  int64  
 9   StoreType                  1017209 non-null  object 
 10  Assortment                 1017209 non-null  object 
 11  CompetitionDistance        1014567 non-null  float64
 12  CompetitionOpenSinceMonth  693861 non-null   float64
 13  CompetitionO

In [11]:
logger.log("test_df.info()")
test_df.info()

2024-09-20 17:56:58,777 - INFO - test_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41088 entries, 0 to 41087
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Id                         41088 non-null  int64  
 1   Store                      41088 non-null  int64  
 2   DayOfWeek                  41088 non-null  int64  
 3   Date                       41088 non-null  object 
 4   Open                       41077 non-null  float64
 5   Promo                      41088 non-null  int64  
 6   StateHoliday               41088 non-null  object 
 7   SchoolHoliday              41088 non-null  int64  
 8   StoreType                  41088 non-null  object 
 9   Assortment                 41088 non-null  object 
 10  CompetitionDistance        40992 non-null  float64
 11  CompetitionOpenSinceMonth  25872 non-null  float64
 12  CompetitionOpenSinceYear   25872 non-null  float64
 13  Promo2                     41088 non-null  int

In [12]:
logger.log("Shape of merged data")
print ('shape of merged train data:' ,train_df.shape)
print ('shape of merged test data:' ,test_df.shape)

2024-09-20 17:59:17,433 - INFO - Shape of merged data


shape of merged train data: (1017209, 18)
shape of merged test data: (41088, 17)


In [14]:
logger.log("summary of statistical metrics for numerical columns of merged train data")
train_df.describe()

2024-09-20 18:08:07,000 - INFO - summary of statistical metrics for numerical columns of merged train data


Store     DayOfWeek         Sales     Customers          Open  \
count  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06   
mean   5.584297e+02  3.998341e+00  5.773819e+03  6.331459e+02  8.301067e-01   
std    3.219087e+02  1.997391e+00  3.849926e+03  4.644117e+02  3.755392e-01   
min    1.000000e+00  1.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    2.800000e+02  2.000000e+00  3.727000e+03  4.050000e+02  1.000000e+00   
50%    5.580000e+02  4.000000e+00  5.744000e+03  6.090000e+02  1.000000e+00   
75%    8.380000e+02  6.000000e+00  7.856000e+03  8.370000e+02  1.000000e+00   
max    1.115000e+03  7.000000e+00  4.155100e+04  7.388000e+03  1.000000e+00   

              Promo  SchoolHoliday  CompetitionDistance  \
count  1.017209e+06   1.017209e+06         1.014567e+06   
mean   3.815145e-01   1.786467e-01         5.430086e+03   
std    4.857586e-01   3.830564e-01         7.715324e+03   
min    0.000000e+00   0.000000e+00         2.000000e+01   
25%    0.000000e+00   0.000000e+00         7.100000e+02   
50%    0.000000e+00   0.000000e+00         2.330000e+03   
75%    1.000000e+00   0.000000e+00         6.890000e+03   
max    1.000000e+00   1.000000e+00         7.586000e+04   

       CompetitionOpenSinceMonth  CompetitionOpenSinceYear        Promo2  \
count              693861.000000             693861.000000  1.017209e+06   
mean                    7.222866               2008.690228  5.005638e-01   
std                     3.211832                  5.992644  4.999999e-01   
min                     1.000000               1900.000000  0.000000e+00   
25%                     4.000000               2006.000000  0.000000e+00   
50%                     8.000000               2010.000000  1.000000e+00   
75%                    10.000000               2013.000000  1.000000e+00   
max                    12.000000               2015.000000  1.000000e+00   

       Promo2SinceWeek  Promo2SinceYear  
count    509178.000000    509178.000000  
mean         23.269093      2011.752774  
std          14.095973         1.662870  
min           1.000000      2009.000000  
25%          13.000000      2011.000000  
50%          22.000000      2012.000000  
75%          37.000000      2013.000000  
max          50.000000      2015.000000

In [15]:
logger.log("summary of statistical metrics for numerical columns of merged test data")
test_df.describe()

2024-09-20 18:08:47,512 - INFO - summary of statistical metrics for numerical columns of merged test data


Id         Store     DayOfWeek          Open         Promo  \
count  41088.000000  41088.000000  41088.000000  41077.000000  41088.000000   
mean   20544.500000    555.899533      3.979167      0.854322      0.395833   
std    11861.228267    320.274496      2.015481      0.352787      0.489035   
min        1.000000      1.000000      1.000000      0.000000      0.000000   
25%    10272.750000    279.750000      2.000000      1.000000      0.000000   
50%    20544.500000    553.500000      4.000000      1.000000      0.000000   
75%    30816.250000    832.250000      6.000000      1.000000      1.000000   
max    41088.000000   1115.000000      7.000000      1.000000      1.000000   

       SchoolHoliday  CompetitionDistance  CompetitionOpenSinceMonth  \
count   41088.000000         40992.000000               25872.000000   
mean        0.443487          5088.583138                   7.035250   
std         0.496802          7225.487467                   3.143015   
min         0.000000            20.000000                   1.000000   
25%         0.000000           720.000000                   4.000000   
50%         0.000000          2425.000000                   7.000000   
75%         1.000000          6480.000000                   9.000000   
max         1.000000         75860.000000                  12.000000   

       CompetitionOpenSinceYear        Promo2  Promo2SinceWeek  \
count              25872.000000  41088.000000     23856.000000   
mean                2008.641929      0.580607        24.426559   
std                    6.862400      0.493466        14.161312   
min                 1900.000000      0.000000         1.000000   
25%                 2006.000000      0.000000        13.000000   
50%                 2010.000000      1.000000        22.000000   
75%                 2012.000000      1.000000        37.000000   
max                 2015.000000      1.000000        49.000000   

       Promo2SinceYear  
count     23856.000000  
mean       2011.820926  
std           1.692166  
min        2009.000000  
25%        2011.000000  
50%        2012.000000  
75%        2013.000000  
max        2015.000000

## Check total null values

In [18]:
logger.log("Total Null values for merged train data")
print(train_df.isnull().sum())

2024-09-20 18:14:35,019 - INFO - Total Null values for merged train data


Store                             0
DayOfWeek                         0
Date                              0
Sales                             0
Customers                         0
Open                              0
Promo                             0
StateHoliday                      0
SchoolHoliday                     0
StoreType                         0
Assortment                        0
CompetitionDistance            2642
CompetitionOpenSinceMonth    323348
CompetitionOpenSinceYear     323348
Promo2                            0
Promo2SinceWeek              508031
Promo2SinceYear              508031
PromoInterval                508031
dtype: int64


In [19]:
logger.log("Total Null values for merged test data")
print(test_df.isnull().sum())

2024-09-20 18:15:01,573 - INFO - Total Null values for merged test data


Id                               0
Store                            0
DayOfWeek                        0
Date                             0
Open                            11
Promo                            0
StateHoliday                     0
SchoolHoliday                    0
StoreType                        0
Assortment                       0
CompetitionDistance             96
CompetitionOpenSinceMonth    15216
CompetitionOpenSinceYear     15216
Promo2                           0
Promo2SinceWeek              17232
Promo2SinceYear              17232
PromoInterval                17232
dtype: int64


In [20]:
logger.log("Calculate the percentage of missing values in each column merged train data")
missing_values = train_df.isnull().sum() / len(train_df)

print(missing_values)

2024-09-20 18:17:01,494 - INFO - Calculate the percentage of missing values in each column merged train data


Store                        0.000000
DayOfWeek                    0.000000
Date                         0.000000
Sales                        0.000000
Customers                    0.000000
Open                         0.000000
Promo                        0.000000
StateHoliday                 0.000000
SchoolHoliday                0.000000
StoreType                    0.000000
Assortment                   0.000000
CompetitionDistance          0.002597
CompetitionOpenSinceMonth    0.317878
CompetitionOpenSinceYear     0.317878
Promo2                       0.000000
Promo2SinceWeek              0.499436
Promo2SinceYear              0.499436
PromoInterval                0.499436
dtype: float64


In [21]:
logger.log("Calculate the percentage of missing values in each column merged test data")
missing_values = test_df.isnull().sum() / len(test_df)

print(missing_values)

2024-09-20 18:17:43,708 - INFO - Calculate the percentage of missing values in each column merged test data


Id                           0.000000
Store                        0.000000
DayOfWeek                    0.000000
Date                         0.000000
Open                         0.000268
Promo                        0.000000
StateHoliday                 0.000000
SchoolHoliday                0.000000
StoreType                    0.000000
Assortment                   0.000000
CompetitionDistance          0.002336
CompetitionOpenSinceMonth    0.370327
CompetitionOpenSinceYear     0.370327
Promo2                       0.000000
Promo2SinceWeek              0.419393
Promo2SinceYear              0.419393
PromoInterval                0.419393
dtype: float64


## Check Duplicated rows

In [22]:
logger.log("Checking for Duplicated Rows")
train_duplicates = train_df.duplicated().sum()
test_duplicates = test_df.duplicated().sum()

# Display the total number of duplicated rows
print(f"Total number of duplicated rows in train data: {train_duplicates}")
print(f"Total number of duplicated rows in test data: {test_duplicates}")


2024-09-20 18:21:00,816 - INFO - Checking for Duplicated Rows


Total number of duplicated rows in train data: 0
Total number of duplicated rows in test data: 0


## Check cardinallity

In [28]:
logger.log("Calculate unique values for object type columns in train data")
train_cardinality = train_df.select_dtypes("object").nunique()
print(train_cardinality)

2024-09-20 18:39:08,142 - INFO - Calculate unique values for object type columns in train data


Date             942
StateHoliday       4
StoreType          4
Assortment         3
PromoInterval      3
dtype: int64


In [26]:
logger.log("Calculate unique values for object type columns in test data")
test_cardinality = test_df.select_dtypes("object").nunique()
print(test_cardinality)

2024-09-20 18:36:08,313 - INFO - Calculate unique values for object type columns in test data


Date             48
StateHoliday      2
StoreType         4
Assortment        3
PromoInterval     3
dtype: int64


---

## General understanding

### 1. First 5 Rows of Merged Train and Test Data
- The **train data** consists of several key columns like `Store`, `Sales`, `Customers`, `Promo`, `StateHoliday`, etc., which capture store performance and promotional activities.
- The **test data** has similar columns but lacks target values like `Sales` and `Customers`.

Example of first 5 rows for train and test data:

| Store | DayOfWeek | Date       | Sales | Customers | Open | Promo | StateHoliday | StoreType | CompetitionDistance |
|-------|-----------|------------|-------|-----------|------|-------|--------------|-----------|----------------------|
| 1     | 5         | 2015-07-31 | 5263  | 555       | 1    | 1     | 0            | c         | 1270.0               |
| 2     | 5         | 2015-07-31 | 6064  | 625       | 1    | 1     | 0            | a         | 570.0                |

### 2. Summary of DataFrame Information
#### Train Data:
- **Total Entries**: 1,017,209
- **Number of Columns**: 18
- Includes features like `Store`, `Sales`, `Promo`, and `CompetitionDistance`.
- Missing data present in columns `CompetitionDistance`, `CompetitionOpenSinceMonth`, `Promo2SinceWeek`, etc.

#### Test Data:
- **Total Entries**: 41,088
- **Number of Columns**: 17
- Similar structure to train data but without `Sales` and `Customers`.

### 3. Statistical Summary of Numerical Columns (Test Data)
| Column                 | Mean     | Std Dev  | Min     | 25%     | 50%     | 75%     | Max     |
|------------------------|----------|----------|---------|---------|---------|---------|---------|
| Store                  | 555.90   | 320.27   | 1.00    | 279.75  | 553.50  | 832.25  | 1115.00 |
| CompetitionDistance     | 5088.58  | 7225.49  | 20.00   | 720.00  | 2425.00 | 6480.00 | 75860.00 |
| Promo2SinceYear         | 2011.82  | 1.69     | 2009.00 | 2011.00 | 2012.00 | 2013.00 | 2015.00 |

### 4. Missing Data Analysis
- **Train Data**: 
  - `CompetitionDistance`: 0.26% missing
  - `CompetitionOpenSinceMonth`: 31.79% missing
  - `Promo2SinceWeek`: 49.94% missing
- **Test Data**: 
  - `CompetitionDistance`: 0.23% missing
  - `Promo2SinceWeek`: 41.93% missing

### 5. Duplicates Check
- **No duplicated rows** were found in both the train and test datasets.

### 6. Unique Values for Object Columns
#### Train Data:
- `Date`: 942 unique values
- `StateHoliday`: 4 unique values
- `StoreType`: 4 unique values
- `Assortment`: 3 unique values
- `PromoInterval`: 3 unique values

#### Test Data:
- `Date`: 48 unique values
- `StateHoliday`: 2 unique values
- `StoreType`: 4 unique values
- `Assortment`: 3 unique values
- `PromoInterval`: 3 unique values

### 7. Data Shape
- **Train Data Shape**: (1,017,209 rows, 18 columns)
- **Test Data Shape**: (41,088 rows, 17 columns)

---

This markdown note summarizes the key points and analysis outcomes of the provided log data effectively.